109 meds loader

Montgomeryyyy opened the pull request:
Implemented a MEDS-loader for the data creation to fix OOM bugs (#149 ). 
Major changes include: 
* Data creation is now done using MEDS data. For this reason logic concerning splitting the data is moved to the MEDS library (MEDS_transform)
* Normalisation has been moved outside of create data, so this is now done in the ehr_preprocess pipeline, which is run before running MEDS_transform (addressing  #123 )
* All dask related function have now been replaced by pandas, and are performed shard-wise addressing the OOM issues in features. 
* Data preparation for training has been moved to a separate script, so that we do not need to run data preparation before each training run. 
* All names have been updated to align with the MEDS names
* All tests have been updated to the new names + MEDS data when necessary (#138 )


All the code has been tested on azure. Currently it takes 2.5-3 hours to create data . Pre-training has however slowed down as I can only use batch sizes of 64, I am not sure why this is so it would be nice if someone could take a look at it. 
Also right now we currently load in all the data before training. This seems to work for now, but we should probably come up with another solution like loading in with the dataloader, as I don't think this will work for long (maybe another issue)?

kirilklein commented on corebehrt/main/finetune_cv.py:
```diff
@@ -32,7 +31,8 @@ def main_finetune(config_path):
     logger = logging.getLogger("finetune_cv")
 
-    data: PatientDataset = DatasetPreparer(cfg).prepare_finetune_data()
+    loaded_data = torch.load(join(cfg.paths.prepared_data, "patients.pt"))
```
Should be a constant

kirilklein commented on corebehrt/main/helper/create_data.py:
```diff
@@ -11,52 +9,79 @@
+    for split_name in ["train", "tuning", "held_out"]:
+        path_name = f"{cfg.paths.data}/{split_name}"
+        if not os.path.exists(path_name):
+            ValueError(f"Path {path_name} does not exist")
```
Should be `raise ValueError`

kirilklein commented on corebehrt/main/helper/create_data.py:
```diff
@@ -11,52 +9,79 @@
+        pids.extend(df["subject_id"].unique().tolist())
+    torch.save(set(pids), join(tokenized_path, f"pids_{split}.pt"))  # save pids as ints
 
 
```
As part of this PR or in a separate issue, making config a bit more specific helps. E.g. instead of passing config, pass features config and paths config. This makes it easier to track what is needed. Ideally, we pass the args directly eg. data_path, feat_path. If we have too many arguments, then it might be a sign that the function needs to be broken down further.

Montgomeryyyy replied:
Yes, I think it makes sense to make this a separate issue. #158

kirilklein commented on corebehrt/modules/features/excluder.py:
```diff
@@ -11,6 +11,6 @@ def __init__(
         self.max_age = max_age
 
-    def exclude_incorrect_events(self, df: dd.DataFrame) -> dd.DataFrame:
+    def exclude_incorrect_events(self, df: pd.DataFrame) -> pd.DataFrame:
```
Since this class is not doing anything except incorrect age exclusion, perhaps we should get rid of it and just use the function?

kirilklein commented on corebehrt/modules/features/features.py:
```diff
@@ -70,27 +60,7 @@ def __call__(
-            """Join concepts with patients_info on PID."""
-            return concepts_partition.merge(patients_info, on="PID", how="left")
+        features = features.drop(columns=["admission_id", "time", "birthdate"])
+        features["subject_id"] = features["subject_id"].astype(str)
```
Why not keep it as int and save some space?

kirilklein commented on corebehrt/modules/features/loader.py:
```diff
@@ -67,87 +63,80 @@ def read_file(file_path: str) -> pd.DataFrame:
+        return df
 
 
 class ParquetIterator:
```
Is this class needed?

Montgomeryyyy replied:
I think it would make sense as we would probably load shards in multiple places with the MEDS format

mikkelfo replied:
I mean since you made a `ShardLoader` will we not always use this rather than `ParquetIterator`?

kirilklein commented on corebehrt/modules/features/values.py:
```diff
@@ -1,114 +1,38 @@
-        values["RESULT"] = (values["RESULT"] * 100).astype(int)
-        values["CONCEPT"] = "VAL_" + values["RESULT"].astype(str)
+    def bin_results(concepts: pd.DataFrame, num_bins=100) -> pd.DataFrame:
+        concepts["numeric_value"] = ValueCreator.bin(
```
This will likely trigger a warning in pandas. Better to make a copy and use .loc

kirilklein commented on setup.py:
```diff
@@ -2,12 +2,13 @@
+    version="0.1",
+    description="",
+    long_description="",
+    authors="Kiril Klein, Mikkel Odgaard, Stephan Lorenzen",
```
+Maria Montgomery :D

mikkelfo commented on corebehrt/constants/data.py:
```diff
@@ -10,22 +10,26 @@
+BIRTHDATE_COL = "birthdate"
+DEATHDATE_COL = "deathdate"
 ABSPOS_COL = "abspos"
+CONCEPT_COL = "code"
```
Fine for now we call it "code", but it's ambiguous if we include values

kirilklein replied:
It's for consistency with MEDS

mikkelfo commented on corebehrt/functional/features/creators.py:
```diff
@@ -152,36 +131,91 @@ def create_segments(concepts: dd.DataFrame) -> dd.DataFrame:
+
+    if admission_indices and discharge_indices:
+        start_idx = 0
+        while start_idx < len(admission_indices):
```
Why is this not a for loop

mikkelfo commented on corebehrt/functional/features/creators.py:
```diff
@@ -152,36 +131,91 @@ def create_segments(concepts: dd.DataFrame) -> dd.DataFrame:
+        start_idx = 0
+        while start_idx < len(admission_indices):
+            start = admission_indices[start_idx]
+            end = next((d for d in discharge_indices if d > start), None)
```
A little freaky to do if our discharge_indices comes from a dataframe, cant we simply do `discharge_indices[discharge_indices > start].index[0]` (or something similar) and drop the conversion to list?

mikkelfo commented on corebehrt/functional/features/creators.py:
```diff
@@ -152,36 +131,91 @@ def create_segments(concepts: dd.DataFrame) -> dd.DataFrame:
+            start_idx += 1
+
+    # Assign admission_id to concepts outside admissions
+    # Concepts within 48 hours of each other are considered to be part of the same admission
```
Is this new functionality or something that was in place?

Montgomeryyyy replied:
As discussed today it has been like this for a while, since many concepts are not in an admission. I have now completely decoupled the admissions and the segments. I feel like that's what we agreed to, otherwise I'll just change it back.

mikkelfo commented on corebehrt/functional/features/exclude.py:
```diff
@@ -1,13 +1,16 @@
+def exclude_event_nans(df: pd.DataFrame, columns=None) -> pd.DataFrame:
     """Exclude events (row) with (any) NaNs"""
-    return df.dropna()
+    if columns:
```
Dont need the if-statement here, since `subset=None` is default, so we can just do `df.dropna(subset=columns)`

mikkelfo commented on corebehrt/functional/features/tokenize.py:
```diff
@@ -22,9 +22,9 @@ def add_special_tokens_partition(
+        cls_rows = df.groupby("subject_id").first()
         # Create [CLS] rows
-        cls_rows["concept"] = CLS_TOKEN
+        cls_rows["code"] = CLS_TOKEN
```
Just leaving this comment here since I've seen it a lot of times, but perhaps we should've just added "code" as a constant and use this constant, rather than having to rename everything to "concept" and "subject_id"

mikkelfo commented on corebehrt/functional/preparation/convert.py:
```diff
@@ -4,6 +4,13 @@
 from tqdm import tqdm
 
 from corebehrt.modules.preparation.dataset import PatientData
+from corebehrt.constants.data import (
```
This file uses `corebehrt.constants` but the others don't - all should use it

mikkelfo commented on corebehrt/main/helper/create_data.py:
```diff
@@ -11,52 +9,79 @@
+        df = pd.read_parquet(shard_path).set_index("subject_id")
+
+        df = tokenizer(df).reset_index()
+        os.makedirs(join(tokenized_path, f"features_{split}"), exist_ok=True)
```
If `tokenized_path` and `split` is pre-defined args, this should be outside the loop?

mikkelfo commented on setup.py:
```diff
@@ -2,12 +2,13 @@
+    version="0.1",
+    description="",
+    long_description="",
+    authors="Kiril Klein, Mikkel Odgaard, Stephan Lorenzen",
```
Should probably add the rest of you or just use a group name instead :)

kirilklein replied:
We should switch to pyproject.toml anyway since the whole setup thing is outdated. But that's also for a separate PR

mikkelfo reviewed (changes requested):
Some minor changes only, but there's two "requests":
1. We need to standardize using the constants, especially in a PR like this where we change "concept" to "code" and "subject_id"
2. We cannot merge this before we have figured out what is going on with the batch size

kirilklein commented on .github/workflows/pipeline.yml:
that's solved

mikkelfo commented:
> For the 2.: are we certain that the memory issue is introduced with this PR?
Yeah there is no issues on main with running larger batch sizes, so it's specific to this PR. Maria and I have been trying to do some debugging to figure out what is going on, but something is weird for sure.

Montgomeryyyy commented on corebehrt/configs/create_data.yaml:
```diff
@@ -3,49 +3,26 @@ logging:
-  #cutoffs:
-   # D: 3 # diagnosis
-   # M: 4 # medication
+  cutoffs:
```
I missed that. Do we already have this, or is this just another issue?

Montgomeryyyy commented on corebehrt/functional/features/creators.py:
```diff
@@ -16,92 +19,69 @@ def create_abspos(concepts: dd.DataFrame, origin_point: datetime) -> dd.DataFram
-
-
-def assign_index_and_order(df: dd.DataFrame) -> dd.DataFrame:
+    concepts.loc[adm_rows.index, "code"] = "ADM_" + concepts.loc[adm_rows.index, "code"]
```
This was basically so that the cutoffs on  the diagnostic codes did not cutoff DISCHARGE. But if we are doing a mapping solution, then this also resolves this problem, and makes it so that we don't have to avoid cutoffs on "A"

Montgomeryyyy commented on corebehrt/configs/prepare_finetune.yaml:
```diff
@@ -0,0 +1,25 @@
+  outcome: ./outputs/outcomes/TEST_OUTCOME.csv
+
+## OUTPUTS
+  prepared_data: ./outputs/finetuning/processed_data/ # Save model/outputs to this folder
```
#159

Montgomeryyyy commented on corebehrt/main/create_data.py:
```diff
@@ -53,7 +50,7 @@ def main_data(config_path):
     # As cfg.paths.features is always set, its value cannot be used to decide
     # if features are present.
-    if os.path.exists(join(cfg.paths.features, "part.0.parquet")):
+    if os.path.exists(join(cfg.paths.features, "held_out", "0.parquet")):
```
I now made the held_out data optional.

Montgomeryyyy commented:
So I think I have now addressed most of the comments. 
I have tried standardising using the constants where I found issues, and also I can run it with larger batch sizes now (not sure why I couldn't before though)

kirilklein reviewed (approved):
Looks good to me!

mikkelfo reviewed (approved):
I'm fine with it as well - would probably just remove the cutoff thing (or comment it out) in the config, just so we don't change files that are not needed
